src/languages/latex.js

```diff
@@ -1,70 +1,243 @@
+    {begin: /[^a-zA-Z@]?/} // control symbol
+  ];
+  const DOUBLE_CARET_VARIANTS = [
+    {begin: /\^{6}[0-9a-f]{6}/},
+    {begin: /\^{5}[0-9a-f]{5}/},
+    {begin: /\^{4}[0-9a-f]{4}/},
+    {begin: /\^{3}[0-9a-f]{3}/},
+    {begin: /\^{2}[0-9a-f]{2}/},
+    {begin: /\^{2}[\u0000-\u007f]/}
+  ];
+  const CONTROL_SEQUENCE = {
+    className: 'keyword',
     begin: /\\/,
     relevance: 0,
     contains: [
       {
-        className: 'name',
-        variants: [
-          {begin: /[a-zA-Z\u0430-\u044f\u0410-\u042f]+[*]?/},
-          {begin: /[^a-zA-Z\u0430-\u044f\u0410-\u042f0-9]/}
-        ],
-        starts: {
-          endsWithParent: true,
-          relevance: 0,
-          contains: [
-            {
-              className: 'string', // because it looks like attributes in HTML tags
-              variants: [
-                {begin: /\[/, end: /\]/},
-                {begin: /\{/, end: /\}/}
-              ]
-            },
-            {
-              begin: /\s*=\s*/, endsWithParent: true,
-              relevance: 0,
-              contains: [
-                {
-                  className: 'number',
-                  begin: /-?\d*\.?\d+(pt|pc|mm|cm|in|dd|cc|ex|em)?/
-                }
-              ]
-            }
-          ]
-        }
+        endsParent: true,
+        begin: KNOWN_CONTROL_WORDS
+      },
+      {
+        endsParent: true,
+        begin: L3_REGEX
+      },
+      {
+        endsParent: true,
+        variants: DOUBLE_CARET_VARIANTS
+      },
+      {
+        endsParent: true,
+        relevance: 0,
+        variants: L2_VARIANTS
       }
     ]
   };
-
-  return {
-    name: 'LaTeX',
-    aliases: ['tex'],
-    contains: [
-      COMMAND,
+  const MACRO_PARAM = {
+    className: 'params',
+    relevance: 0,
+    begin: /#+\d?/
+  };
+  const DOUBLE_CARET_CHAR = {
+    // relevance: 1
+    variants: DOUBLE_CARET_VARIANTS
+  };
+  const SPECIAL_CATCODE = {
+    className: 'built_in',
+    relevance: 0,
+    begin: /[$&^_]/
+  };
+  const MAGIC_COMMENT = {
+    className: 'meta',
+    begin: '% !TeX',
+    end: '$',
+    relevance: 10
+  };
+  const COMMENT = hljs.COMMENT(
+    '%',
+    '$',
+    {
+      relevance: 0
+    }
+  );
+  const EVERYTHING_BUT_VERBATIM = [
+    CONTROL_SEQUENCE,
+    MACRO_PARAM,
+    DOUBLE_CARET_CHAR,
+    SPECIAL_CATCODE,
+    MAGIC_COMMENT,
+    COMMENT
+  ];
+  const BRACE_GROUP_NO_VERBATIM = {
+    begin: /\{/, end: /\}/,
+    relevance: 0,
+    contains: ['self', ...EVERYTHING_BUT_VERBATIM]
+  };
+  const ARGUMENT_BRACES = hljs.inherit(
+    BRACE_GROUP_NO_VERBATIM,
+    {
+      relevance: 0,
+      endsParent: true,
+      contains: [BRACE_GROUP_NO_VERBATIM, ...EVERYTHING_BUT_VERBATIM]
+    }
+  );
+  const ARGUMENT_BRACKETS = {
+    begin: /\[/,
+      end: /\]/,
+    endsParent: true,
+    relevance: 0,
+    contains: [BRACE_GROUP_NO_VERBATIM, ...EVERYTHING_BUT_VERBATIM]
+  };
+  const END_PARENT = {
+    begin: /(?=[.$])/,
+    end: /(?=[.$])/,
```

Comment:
Explain again please, is this meant to be `.` (any character) or the literal `.` (period).

Can you show me an example or two that this is intended to terminate?

Reply:
It's any character. This rule is meant to give a zero-length match anywhere (looking ahead for any character or the end of the line).

This is used when looking for optional arguments. The grammar is meant to look for a `[`. If it finds one, it should highlight the entire optional argument and end the parent (i.e. stop looking for optional arguments). If it doesn't find one, it should end the parent anyway, instead of stepping on to the next character and trying again. That is what this mode does: End the parent, no matter what. (Without consuming any characters.)

Reply:
Are you sure it works?  `.` inside `[]` is a literal period

```
> "test\nbooger".match(/(?=[.$])/)
null
> "test\nbooger".match(/(?=.|$)/)
[ '', index: 0, input: 'test\nbooger', groups: undefined ]
> "test\nsmith.\nbooger".match(/(?=[.$])/)
[ '', index: 10, input: 'test\nsmith.\nbooger', groups: undefined ]
```

And if I'm correct and it's not even working, is it necessary?  Where is the failing test?

Reply:
Any FYI our "end immediately" goto regex is `/\B|\b/`

Reply:
I presume something incorrect could get highlighted (if the parent mode continued) but our tests seem to indicate that most often the parent mode is terminating on it's own (making me wonder if we really need this in actuality)... if we do we may need to revisit it later (you could open an issue) since a correct rule is going to run afoul of the fact that rules are not allowed to consume nothing.  I'm not sure it should hold up the merge unless it's a major issue.

Reply:
Hmm, I thought I tested its necessity. Give me a minute..

Reply:
I couldn't produce any problems without that mode, so I removed it. Seems to have been a chimera..

Reply:
Modes without an `end` will get the default `end: /\B|\b/` so they are very eager to end themselves UNLESS they continue to match children... so the first time no children can be found at the current cursor position the parent mode terminates.  `end` on the other hand "holds the mode open" and allows for non-matches in-between the children matches.

This makes it hard for us to express:

- mode A followed by a single mode B

Because if there are multiple B's immediately in a row under A, then they will continue to get matched... `endsParent` solves this for single paired items but gets complex when you want to go deeper.